modules/tensor_mechanics/include/userobjects/EulerAngleProvider.h

```diff
@@ -28,6 +28,8 @@ class EulerAngleProvider : public GeneralUserObject
   EulerAngleProvider(const InputParameters & parameters) : GeneralUserObject(parameters) {}
 
   virtual const EulerAngles & getEulerAngles(unsigned int) const = 0;
+  virtual unsigned int getGlobalID(unsigned int phase, unsigned int local_id) const = 0;
+  virtual unsigned int getGrainNum(unsigned int phase) const = 0;
```

Comment:
Adding the new virtual here looks ok to me. I think my only issue is that you probably shouldn't make it pure. We don't want to limit ourselves with these generic classes and force something like phase on everything downstream. What you should do is just provide a default that throws and error message if you can't provide a good default at the base class level. This would allow users to still use this class for single phase problems without changing anything.

Also, sorry for the delay. Several people were out during December including myself.

Reply:
Thanks for the response.  The original code (before my changes) used a virtual pure method for "getGrainNum" that is also used in the downstream objects.  My first pull request added a second "getGrainNum" object that was phase dependent but not pure. From the previous comments, my understanding was that I should change it to pure and add code to the downstream objects so we didn't have mixed objects.  I did that with my second pull request.  To clarify, are you saying I should change it back so one is pure and the other not?

Reply:
Daniel is right, this interface still isn't in the right place. However, to address your comment specifically, what I was looking for was something in between. Instead of adding a new pure virtual, I was suggesting that you add a new virtual but not implement it in the class (instead you'd throw and error if the method was called on the base class). Why do it this way? Well it gives you the ability to add a new method that "should" be implemented but isn't required all the way down.

You can tell that this isn't in the right place since your new data structure has a comment that says something to the effect that "this isn't needed here". 

Perhaps, @dschwen, myself and you can all chat about the right design here. I believe we are both in the office this week.

Reply:
@frombs - Let's switch this so that it's not pure. Just provide an error message in the base class if you can't provide a concrete implementation. That way you only have to implement this where you will use it, not everywhere.